File: web/src/stores/blueprint.ts
import {defineStore} from 'pinia'
import {computed, ref} from 'vue'
import type {Blueprint, Function, Connection, Node, Position, Variable} from '../types/blueprint'
import {v4 as uuid} from 'uuid'
import {useWorkspaceStore} from "./workspace";
import {isEqual} from 'lodash'
import {EventDefinition} from "../services/eventService"; // Need to add this import for comparing objects

export const useBlueprintStore = defineStore('blueprint', () => {
    // State
    const blueprint = ref<Blueprint>({
        id: '',
        name: '',
        description: '',
        version: '1.0',
        functions: [],
        nodes: [],
        connections: [],
        variables: [],
        events: [],
        eventBindings: [],
        metadata: {}
    })

    // Track the latest saved state of the blueprint for comparison
    const latestSavedBlueprint = ref<Blueprint | null>(null)
    
    const isLoading = ref(false)
    const error = ref<string | null>(null)

    const currentEditingFunction = ref<string | null>(null)
    
    // Track available versions for the current blueprint
    const availableVersions = ref<{versionNumber: number, createdAt: string, comment: string}[]>([])

    // Getters
    const nodes = computed(() => blueprint.value.nodes || [])
    const connections = computed(() => blueprint.value.connections || [])
    const variables = computed(() => {
        if (!blueprint.value.variables) {
            blueprint.value.variables = [];
        }
        return blueprint.value.variables;
    })
    const functions = computed(() => blueprint.value.functions || [])

    // Check if there are unsaved changes by comparing with the last saved state
    const hasUnsavedChanges = computed(() => {
        if (!latestSavedBlueprint.value) return true;
        
        // Perform a deep comparison between current blueprint and latest saved
        return !isEqual(blueprint.value, latestSavedBlueprint.value);
    })

    const getNodeById = computed(() => (id: string) => {
        return blueprint.value.nodes.find(node => node.id === id)
    })

    const getNodeConnections = computed(() => (nodeId: string) => {
        return blueprint.value.connections.filter(
            conn => conn.sourceNodeId === nodeId || conn.targetNodeId === nodeId
        )
    })

    const getNodeInputConnections = computed(() => (nodeId: string) => {
        return blueprint.value.connections.filter(conn => conn.targetNodeId === nodeId)
    })

    const getNodeOutputConnections = computed(() => (nodeId: string) => {
        return blueprint.value.connections.filter(conn => conn.sourceNodeId === nodeId)
    })

    const isNodePinConnected = computed(() => (nodeId: string, pinId: string, direction: 'input' | 'output') => {
        if (direction === 'input') {
            return blueprint.value.connections.some(
                conn => conn.targetNodeId === nodeId && conn.targetPinId === pinId
            )
        } else {
            return blueprint.value.connections.some(
                conn => conn.sourceNodeId === nodeId && conn.sourcePinId === pinId
            )
        }
    })

    const findEntryPoints = computed(() => {
        const execInputs = new Set<string>()
        const execOutputs = new Set<string>()

        // Find all nodes with execution connections
        for (const conn of blueprint.value.connections) {
            if (conn.connectionType === 'execution') {
                execOutputs.add(conn.sourceNodeId)
                execInputs.add(conn.targetNodeId)
            }
        }

        // Find nodes with execution outputs but no execution inputs
        const entryPoints = blueprint.value.nodes
            .filter(node => execOutputs.has(node.id) && !execInputs.has(node.id))
            .map(node => node.id)

        // Also include special entry point nodes like DOM events
        blueprint.value.nodes.forEach(node => {
            if (node.type === 'dom-event' && !entryPoints.includes(node.id)) {
                entryPoints.push(node.id)
            }
        })

        return entryPoints
    })

    const isFunctionEditing = computed(() => currentEditingFunction.value !== null)

    // Actions
    function createBlueprint(name: string, description: string = '') {
        blueprint.value = {
            id: uuid(),
            name,
            description,
            version: '1.0',
            functions: [],
            nodes: [],
            connections: [],
            variables: [],
            events: [],
            eventBindings: [],
            metadata: {}
        }
        // Reset the saved state
        latestSavedBlueprint.value = null;
        // Reset available versions
        availableVersions.value = [];
    }

    async function loadBlueprint(id: string) {
        isLoading.value = true
        error.value = null

        try {
            const response = await fetch(`/api/blueprints/${id}`)
            if (!response.ok) {
                throw new Error(`Failed to load blueprint: ${response.statusText}`)
            }

            blueprint.value = await response.json()
            
            // Ensure the blueprint structure is complete
            if (!blueprint.value.variables) {
                console.log('Variables array not found in blueprint response, creating empty array');
                blueprint.value.variables = [];
            }
            
            if (!blueprint.value.functions) {
                blueprint.value.functions = [];
            }
            
            if (!blueprint.value.connections) {
                blueprint.value.connections = [];
            }
            
            if (!blueprint.value.nodes) {
                blueprint.value.nodes = [];
            }

            if (!blueprint.value.events) {
                blueprint.value.events = [];
            }

            if (!blueprint.value.eventBindings) {
                blueprint.value.eventBindings = [];
            }

            console.log('Blueprint structure after loading:', blueprint.value);
            
            // Save the loaded blueprint as the latest saved state
            latestSavedBlueprint.value = JSON.parse(JSON.stringify(blueprint.value));
            
            // Load available versions
            await loadBlueprintVersions(id);
        } catch (err) {
            error.value = err instanceof Error ? err.message : String(err)
            console.error('Error loading blueprint:', err)
        } finally {
            isLoading.value = false
        }
    }

    async function loadBlueprintVersions(blueprintId: string) {
        try {
            const response = await fetch(`/api/blueprints/${blueprintId}/versions`);
            if (!response.ok) {
                throw new Error(`Failed to load blueprint versions: ${response.statusText}`);
            }
            
            availableVersions.value = await response.json();
            console.log('Loaded versions:', availableVersions.value);
        } catch (err) {
            console.error('Error loading blueprint versions:', err);
        }
    }

    async function loadBlueprintVersion(blueprintId: string, versionNumber: number) {
        isLoading.value = true;
        error.value = null;

        try {
            const response = await fetch(`/api/blueprints/${blueprintId}/versions/${versionNumber}`);
            if (!response.ok) {
                throw new Error(`Failed to load blueprint version: ${response.statusText}`);
            }

            blueprint.value = await response.json();
            
            // Ensure the blueprint structure is complete
            if (!blueprint.value.variables) {
                blueprint.value.variables = [];
            }
            
            if (!blueprint.value.functions) {
                blueprint.value.functions = [];
            }
            
            if (!blueprint.value.connections) {
                blueprint.value.connections = [];
            }
            
            if (!blueprint.value.nodes) {
                blueprint.value.nodes = [];
            }
            
            // Update the latest saved blueprint to match this version
            latestSavedBlueprint.value = JSON.parse(JSON.stringify(blueprint.value));
            
            console.log(`Loaded blueprint version ${versionNumber}`);
        } catch (err) {
            error.value = err instanceof Error ? err.message : String(err);
            console.error('Error loading blueprint version:', err);
        } finally {
            isLoading.value = false;
        }
    }

    async function saveBlueprint(workspaceId?: string) {
        isLoading.value = true;
        error.value = null;

        try {
            // If there are no unsaved changes, skip saving
            if (latestSavedBlueprint.value && !hasUnsavedChanges.value) {
                console.log('No changes detected, skipping save operation');
                isLoading.value = false;
                return blueprint.value;
            }

            // Get workspace ID if not provided
            const workspace = workspaceId || useWorkspaceStore().currentWorkspace?.id;
            if (!workspace) {
                throw new Error('No workspace specified for saving blueprint');
            }

            const method = blueprint.value.id ? 'PUT' : 'POST';
            const url = blueprint.value.id
                ? `/api/blueprints/${blueprint.value.id}?workspace=${workspace}`
                : `/api/blueprints?workspace=${workspace}`;

            const response = await fetch(url, {
                method,
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(blueprint.value)
            });

            if (!response.ok) {
                throw new Error(`Failed to save blueprint: ${response.statusText}`);
            }

            blueprint.value = await response.json();
            
            // Ensure the blueprint structure is complete after saving
            if (!blueprint.value.variables) {
                console.log('Variables array not found in blueprint response after saving, creating empty array');
                blueprint.value.variables = [];
            }
            
            if (!blueprint.value.functions) {
                blueprint.value.functions = [];
            }
            
            if (!blueprint.value.connections) {
                blueprint.value.connections = [];
            }
            
            if (!blueprint.value.nodes) {
                blueprint.value.nodes = [];
            }

            if (!blueprint.value.eventBindings) {
                blueprint.value.eventBindings = [];
            }
            
            // Update the latest saved state
            latestSavedBlueprint.value = JSON.parse(JSON.stringify(blueprint.value));
            
            // Refresh available versions
            await loadBlueprintVersions(blueprint.value.id);
            
            return blueprint.value;
        } catch (err) {
            error.value = err instanceof Error ? err.message : String(err);
            console.error('Error saving blueprint:', err);
            throw err;
        } finally {
            isLoading.value = false;
        }
    }

    // Function to manually create a new version with a comment
    async function createNewVersion(comment: string = 'Manual save') {
        if (!blueprint.value.id) {
            error.value = 'Cannot create a version for an unsaved blueprint';
            return;
        }

        isLoading.value = true;
        error.value = null;

        try {
            const response = await fetch(`/api/blueprints/${blueprint.value.id}/versions`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({ comment })
            });

            if (!response.ok) {
                throw new Error(`Failed to create blueprint version: ${response.statusText}`);
            }

            const result = await response.json();
            console.log('Created new version:', result);
            
            // Refresh available versions
            await loadBlueprintVersions(blueprint.value.id);
            
            // Update the latest saved state
            latestSavedBlueprint.value = JSON.parse(JSON.stringify(blueprint.value));
            
            return result.versionNumber;
        } catch (err) {
            error.value = err instanceof Error ? err.message : String(err);
            console.error('Error creating blueprint version:', err);
            throw err;
        } finally {
            isLoading.value = false;
        }
    }

    function checkNodeBind(node: Node) {
        return node.type === 'event-bind'
    }

    function addNode(node: Node) {
        // Make sure we're creating a deep copy to avoid shared references
        const nodeCopy = JSON.parse(JSON.stringify(node));

        // Ensure properties is initialized
        if (!nodeCopy.properties) {
            nodeCopy.properties = [];
        }

        // Add the node
        blueprint.value.nodes.push(nodeCopy);
    }

    function updateNode(id: string, updates: Partial<Node>) {
        const index = blueprint.value.nodes.findIndex(node => node.id === id)
        if (index !== -1) {
            blueprint.value.nodes[index] = { ...blueprint.value.nodes[index], ...updates }
        }
    }

    function updateNodePosition(id: string, position: Position) {
        const node = blueprint.value.nodes.find(node => node.id === id)
        if (node) {
            node.position = position
        }
    }

    function updateNodeProperty(nodeId: string, propertyName: string, value: any) {
        const node = blueprint.value.nodes.find(node => node.id === nodeId)
        if (node) {
            // Find the property
            const propIndex = node.properties?.findIndex(p => p.name === propertyName) ?? -1
            if (propIndex !== -1) {
                // Update existing property
                node.properties[propIndex].value = value
            } else {
                // Add new property
                if (!node.properties) {
                    node.properties = [];
                }
                node.properties.push({ displayName: node.properties[propIndex]?.displayName ?? propertyName, name: propertyName, value: value })
            }

            // If this is a pin default value (input_*), add it to the node's data for easy access
            if (propertyName.startsWith('input_')) {
                if (!node.data) {
                    node.data = {}
                }
                if (!node.data.defaults) {
                    node.data.defaults = {}
                }
                const pinId = propertyName.substring(6) // Remove "input_" prefix
                node.data.defaults[pinId] = value
            }
        }
    }

    function removeNode(id: string) {
        // Remove connections first
        blueprint.value.connections = blueprint.value.connections.filter(
            conn => conn.sourceNodeId !== id && conn.targetNodeId !== id
        )

        // Remove node
        blueprint.value.nodes = blueprint.value.nodes.filter(node => node.id !== id)
    }

    function findNode(id: string) {
        return blueprint.value.nodes.find((n: Node) => n.id === id)
    }

    function addConnection(connection: Connection) {
        // Ensure the connection has an ID
        if (!connection.id) {
            connection.id = uuid()
        }

        // Check if a similar connection already exists
        const exists = blueprint.value.connections.some(
            conn =>
                conn.sourceNodeId === connection.sourceNodeId &&
                conn.sourcePinId === connection.sourcePinId &&
                conn.targetNodeId === connection.targetNodeId &&
                conn.targetPinId === connection.targetPinId
        )

        const node = findNode(connection.sourceNodeId);
        if (checkNodeBind(node)) {
            addEventBinding(node);
        }

        if (!exists) {
            blueprint.value.connections.push(connection)
        }
    }

    function removeConnection(id: string) {
        blueprint.value.connections = blueprint.value.connections.filter(
            conn => conn.id !== id
        )
    }

    async function addVariable(workspaceId: string, variable: Variable) {
        // [sink] try to save blueprint before saving variable
        await saveBlueprint(workspaceId)

        try {
            const url = `/api/blueprints/${blueprint.value.id}/variable`
            const variableResponse = await fetch(url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    ...variable,
                    version: blueprint.value.version
                }),
            })

            if (!variableResponse.ok) {
                throw new Error('Failed to add variable into blueprint')
            }
            // Ensure variables array exists
            if (!blueprint.value.variables) {
                console.log('Creating variables array before adding variable');
                blueprint.value.variables = [];
            }

            // Add the variable
            console.log('Adding variable to blueprint:', variable);
            blueprint.value.variables.push(await variableResponse.json());
            
            // Update the latest saved state
            latestSavedBlueprint.value = JSON.parse(JSON.stringify(blueprint.value));
        } catch (e) {
            console.log('Something went wrong while adding variable to blueprint', e)
        }
    }

    function addFunction(fn: Function) {
        // Ensure function has execution pins
        let hasExecutionInput = false;
        let hasExecutionOutput = false;
        
        for (const input of fn.nodeType.inputs) {
            if (input.type?.id === 'execution') {
                hasExecutionInput = true;
                break;
            }
        }
        
        for (const output of fn.nodeType.outputs) {
            if (output.type?.id === 'execution') {
                hasExecutionOutput = true;
                break;
            }
        }
        
        // Add default execution input if needed
        if (!hasExecutionInput) {
            fn.nodeType.inputs.push({
                id: 'exec',
                name: 'Execute',
                description: 'Execution continues',
                type: {
                    id: 'execution',
                    name: 'Execution',
                    description: 'Controls execution flow',
                },
                optional: false
            });
        }
        
        // Add default execution output if needed
        if (!hasExecutionOutput) {
            fn.nodeType.outputs.push({
                id: 'then',
                name: 'Then',
                description: 'Execution continues',
                type: {
                    id: 'execution',
                    name: 'Execution',
                    description: 'Controls execution flow',
                },
                optional: false
            });
        }
        
        blueprint.value.functions.push(fn)
    }

    function addNodeToFunction(fnID: string, node: Node) {
        const fnIdx = blueprint.value.functions.findIndex((fn: Function) => fn.id === fnID)
        blueprint.value.functions[fnIdx].nodes.push(node)
    }

    function addEvent(e: EventDefinition) {
        if (!blueprint.value.events) {
            blueprint.value.events = [];
        }

        blueprint.value.events.push(e)
    }

    function findEvent(id: string) {
        return blueprint.value.events.find((e: EventDefinition) => e.id === id)
    }

    function addEventBinding(node: Node) {
        if (!blueprint.value.eventBindings) {
            blueprint.value.eventBindings = [];
        }

        let eventId = '';
        let priority = 0;
        node.properties.forEach((prop) => {
            if (prop.name === 'eventID') {
                eventId = prop.value;
            }

            if (prop.name === 'priority') {
                priority = prop.value;
            }
        });

        const event = findEvent(eventId);
        if (!event) {
            return
        }

        blueprint.value.eventBindings.push({
            id: `binding.${eventId}.${node.id}`,
            eventId: eventId,
            handlerId: node.id,
            handlerType: node.type,
            blueprintId: blueprint.value.id,
            priority: priority,
            enabled: true,
            createdAt: Date.now().toString(),
        })
    }

    return {
        blueprint,
        isLoading,
        error,
        currentEditingFunction,
        nodes,
        connections,
        getNodeById,
        getNodeConnections,
        getNodeInputConnections,
        getNodeOutputConnections,
        isNodePinConnected,
        isFunctionEditing,
        findEntryPoints,
        hasUnsavedChanges,
        availableVersions,
        createBlueprint,
        loadBlueprint,
        loadBlueprintVersion,
        loadBlueprintVersions,
        saveBlueprint,
        createNewVersion,
        addNode,
        updateNode,
        updateNodePosition,
        updateNodeProperty,
        removeNode,
        addConnection,
        removeConnection,
        variables,
        addVariable,
        functions,
        addFunction,
        addNodeToFunction,
        addEvent,
        addEventBinding,
    }
})

File: pkg/blueprint/blueprint.go
package blueprint

// BlueprintNode represents a node in a blueprint
type BlueprintNode struct {
	ID         string                 `json:"id"`
	Type       string                 `json:"type"`
	Position   Position               `json:"position"`
	Properties []NodeProperty         `json:"properties"`
	Data       map[string]interface{} `json:"data,omitempty"`
}

type BlueprintNodeType struct {
	Inputs     []NodePin      `json:"inputs"`
	Outputs    []NodePin      `json:"outputs"`
	Properties []NodeProperty `json:"properties"`
}

type NodePin struct {
	ID          string       `json:"id,omitempty"`          // Unique identifier
	Name        string       `json:"name,omitempty"`        // Human-readable name
	Description string       `json:"description,omitempty"` // Description of what the pin does
	Type        *NodePinType `json:"type,omitempty"`        // Type of data for this pin
	Optional    bool         `json:"optional,omitempty"`    // Whether this pin is required
	Default     interface{}  `json:"default,omitempty"`     // Default value if not connected
}

type NodePinType struct {
	ID          string `json:"id,omitempty"`
	Name        string `json:"name,omitempty"`
	Description string `json:"description,omitempty"`
}

// Position represents the node position on the canvas
type Position struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

// NodeProperty represents a property of a node
type NodeProperty struct {
	Name        string       `json:"name"`
	DisplayName string       `json:"displayName"`
	Description string       `json:"description"`
	Value       interface{}  `json:"value"`
	Type        *NodePinType `json:"type"`
}

// Connection represents a connection between nodes
type Connection struct {
	ID             string         `json:"id"`
	SourceNodeID   string         `json:"sourceNodeId"`
	SourcePinID    string         `json:"sourcePinId"`
	TargetNodeID   string         `json:"targetNodeId"`
	TargetPinID    string         `json:"targetPinId"`
	ConnectionType string         `json:"connectionType"` // "execution" or "data"
	Data           map[string]any `json:"data"`
}

// Variable represents a blueprint variable
type Variable struct {
	Name  string      `json:"name"`
	Type  string      `json:"type"`
	Value interface{} `json:"value"`
}

type Function struct {
	ID          string            `json:"id"`
	Name        string            `json:"name"`
	Description string            `json:"description,omitempty"`
	NodeType    BlueprintNodeType `json:"nodeType"`
	Nodes       []BlueprintNode   `json:"nodes"`
	Connections []Connection      `json:"connections"`
	Variables   []Variable        `json:"variables,omitempty"`
	Metadata    map[string]string `json:"metadata,omitempty"`
}

// Blueprint represents a complete blueprint definition
type Blueprint struct {
	ID            string            `json:"id"`
	Name          string            `json:"name"`
	Description   string            `json:"description,omitempty"`
	Version       string            `json:"version"`
	Nodes         []BlueprintNode   `json:"nodes"`
	Functions     []Function        `json:"functions"`
	Connections   []Connection      `json:"connections"`
	Variables     []Variable        `json:"variables,omitempty"`
	Metadata      map[string]string `json:"metadata,omitempty"`
	Events        []EventDefinition `json:"events,omitempty"`         // User-defined event definitions
	EventBindings []EventBinding    `json:"event_bindings,omitempty"` // User-defined event bindings
}

// EventParameter defines a parameter for a custom event within a blueprint
type EventParameter struct {
	Name        string      `json:"name"`                  // Parameter name
	TypeID      string      `json:"typeId"`                // Pin Type ID (e.g., "string", "number")
	Description string      `json:"description,omitempty"` // Description of the parameter
	Optional    bool        `json:"optional,omitempty"`    // Whether the parameter is optional
	Default     interface{} `json:"default,omitempty"`     // Default value for optional parameters
}

// EventDefinition defines a custom event within a blueprint
type EventDefinition struct {
	ID          string           `json:"id"`                    // Unique identifier for the event (e.g., "custom.my-event")
	Name        string           `json:"name"`                  // Human-readable name
	Description string           `json:"description,omitempty"` // Description of what the event does
	Parameters  []EventParameter `json:"parameters,omitempty"`  // Parameters that can be passed with the event
	Category    string           `json:"category,omitempty"`    // Category for organization (defaults to "Custom")
}

// EventBinding defines a binding between an event and a handler
type EventBinding struct {
	ID          string `json:"id"`           // Unique identifier for the binding
	EventID     string `json:"event_id"`     // ID of the event to bind to
	HandlerID   string `json:"handler_id"`   // ID of the node that handles the event
	HandlerType string `json:"handler_type"` // Type of handler (e.g., "node", "function")
	Priority    int    `json:"priority"`     // Priority for execution order
	Enabled     bool   `json:"enabled"`      // Whether the binding is enabled
}

// NewBlueprint creates a new empty blueprint
func NewBlueprint(id, name, version string) *Blueprint {
	return &Blueprint{
		ID:            id,
		Name:          name,
		Version:       version,
		Nodes:         make([]BlueprintNode, 0),
		Connections:   make([]Connection, 0),
		Variables:     make([]Variable, 0),
		Metadata:      make(map[string]string),
		Events:        make([]EventDefinition, 0), // Initialize Events slice
		EventBindings: make([]EventBinding, 0),    // Initialize EventBindings slice
	}
}

// AddNode adds a node to the blueprint
func (b *Blueprint) AddNode(node BlueprintNode) {
	b.Nodes = append(b.Nodes, node)
}

// AddConnection adds a connection to the blueprint
func (b *Blueprint) AddConnection(conn Connection) {
	b.Connections = append(b.Connections, conn)
}

// AddVariable adds a variable to the blueprint
func (b *Blueprint) AddVariable(variable Variable) {
	b.Variables = append(b.Variables, variable)
}

// FindNode finds a node by ID
func (b *Blueprint) FindNode(id string) *BlueprintNode {
	for i := range b.Nodes {
		if b.Nodes[i].ID == id {
			return &b.Nodes[i]
		}
	}
	return nil
}

// FindConnection finds a connection by ID
func (b *Blueprint) FindConnection(id string) *Connection {
	for i := range b.Connections {
		if b.Connections[i].ID == id {
			return &b.Connections[i]
		}
	}
	return nil
}

// FindVariable finds a variable by name
func (b *Blueprint) FindVariable(name string) *Variable {
	for i := range b.Variables {
		if b.Variables[i].Name == name {
			return &b.Variables[i]
		}
	}
	return nil
}

// GetNodeConnections finds all connections for a node
func (b *Blueprint) GetNodeConnections(nodeID string) []Connection {
	connections := make([]Connection, 0)

	for _, conn := range b.Connections {
		if conn.SourceNodeID == nodeID || conn.TargetNodeID == nodeID {
			connections = append(connections, conn)
		}
	}

	return connections
}

// GetNodeInputConnections finds all input connections for a node
func (b *Blueprint) GetNodeInputConnections(nodeID string) []Connection {
	connections := make([]Connection, 0)

	for _, conn := range b.Connections {
		if conn.TargetNodeID == nodeID {
			connections = append(connections, conn)
		}
	}

	return connections
}

// GetNodeOutputConnections finds all output connections for a node
func (b *Blueprint) GetNodeOutputConnections(nodeID string) []Connection {
	connections := make([]Connection, 0)

	for _, conn := range b.Connections {
		if conn.SourceNodeID == nodeID {
			connections = append(connections, conn)
		}
	}

	return connections
}

// RemoveNode removes a node and all its connections
func (b *Blueprint) RemoveNode(nodeID string) {
	// Remove connections first
	newConnections := make([]Connection, 0)
	for _, conn := range b.Connections {
		if conn.SourceNodeID != nodeID && conn.TargetNodeID != nodeID {
			newConnections = append(newConnections, conn)
		}
	}
	b.Connections = newConnections

	// Remove node
	newNodes := make([]BlueprintNode, 0)
	for _, node := range b.Nodes {
		if node.ID != nodeID {
			newNodes = append(newNodes, node)
		}
	}
	b.Nodes = newNodes
}

// RemoveConnection removes a connection
func (b *Blueprint) RemoveConnection(connID string) {
	newConnections := make([]Connection, 0)
	for _, conn := range b.Connections {
		if conn.ID != connID {
			newConnections = append(newConnections, conn)
		}
	}
	b.Connections = newConnections
}

// FindEntryPoints finds nodes that should be triggered first
// (nodes with execution outputs but no execution inputs)
func (b *Blueprint) FindEntryPoints() []string {
	entryPoints := make([]string, 0)

	//// Create maps to track nodes with execution inputs and outputs
	//execInputs := make(map[string]bool)
	//execOutputs := make(map[string]bool)
	//
	//// Find all nodes with execution connections
	//for _, conn := range b.Connections {
	//	if conn.ConnectionType == "execution" {
	//		execOutputs[conn.SourceNodeID] = true
	//		execInputs[conn.TargetNodeID] = true
	//	}
	//}
	//
	//// Find nodes with execution outputs but no execution inputs
	//for _, node := range b.Nodes {
	//	if execOutputs[node.ID] && !execInputs[node.ID] {
	//		entryPoints = append(entryPoints, node.ID)
	//	}
	//}

	// Also include special entry point nodes like DOM events
	for _, node := range b.Nodes {
		for _, property := range node.Properties {
			if property.Name == "eventType" && property.Value.(string) == "entry" {
				entryPoints = append(entryPoints, node.ID)
			}
		}
	}

	return entryPoints
}

// Helper function to check if a slice contains a string
func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}
